app/controllers/authentications/microsoft_graph_oauth_controller.rb

```diff
@@ -0,0 +1,96 @@
+class Authentications::MicrosoftGraphOauthController < ApplicationController
+  allow_unauthenticated_access
+
+  def create
+    if Current.user
+      Current.user.microsoft_graph_credential&.destroy
+      _, cred = add_person_credentials("MicrosoftGraphCredential")
+      cred.save! && redirect_to(edit_settings_person_path, notice: "Saved") && return
+
+    elsif (credential = MicrosoftGraphCredential.find_by(oauth_id: auth[:uid]))
+      @person = credential.user.person
+
+    elsif Feature.disabled?(:registration)
+      redirect_to(root_path, alert: "Registration is disabled") && return
+
+    elsif auth_email && (user = Person.find_by(email: auth_email)&.user)
+      @person = init_for_user(user)
+
+    elsif auth_email && (@person = Person.find_by(email: auth_email))
+      @person = init_for_person(@person)
+
+    else
+      @person = initialize_microsoft_person
+    end
+
+    if @person&.save
+      login_as(@person, credential: @person.user.reload.microsoft_graph_credential)
+      redirect_to root_path
+    else
+      @person&.errors&.delete :personable
+      msg = @person.errors.full_messages.map { |m| m.gsub(/Personable |credentials /, "") }.to_sentence.capitalize
+      if msg.downcase.include?("oauth refresh token can't be blank")
+        msg += " " + helpers.link_to("Microsoft third-party connections", "https://microsoft.com/consent", class: "underline") + " search for website, and delete all it's connections. Then try again."
```

Comment:
Are you sure this link shouldn't be https://account.microsoft.com/privacy/app-access ? That's what you put in the READMe. I'm not sure which one is the easier path for a user to take.

Reply:
For anyone interested here is where the two links go, similar but different.

1. https://account.microsoft.com/privacy/app-access
2. https://microsoft.com/consent

<img width="640" alt="Screenshot 2024-11-14 at 6 52 15 am" src="https://github.com/user-attachments/assets/f79e5caf-8805-415f-bf04-6c169aa51ffd">

The suggested former one looks better; will switch.